File: library/src/main/java/com/handy/basic/app/BaseFragment.java
package com.handy.basic.app;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;

import com.blankj.utilcode.util.LogUtils;
import com.blankj.utilcode.util.ScreenUtils;
import com.handy.basic.evenbus.BaseMessageEvent;
import com.trello.rxlifecycle2.components.support.RxFragment;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

/**
 * BaseFragment
 *
 * @author LiuJie https://github.com/Handy045
 * @description functional description.
 * @date Created in 2019/2/27 16:53
 * @modified By liujie
 */
public abstract class BaseFragment extends RxFragment implements BaseApplicationApi.BaseFragmentApi {

    //============================================================
    //  功能配置
    //============================================================

    /**
     * Fragment是否已创建成功，用于控制在Fragment创建之前不调用 {@link BaseFragment#setUserVisibleHint(boolean)} 方法实现内部的 {@link BaseFragment#onVisiableHDB()} ()} 和 {@link BaseFragment#onLazyLoadHDB()} 方法
     */
    private boolean isCreateed = false;
    /**
     * onViewCreated中初始化界面视图
     */
    private boolean isInitViewHDB = true;
    /**
     * onActivityCreated中初始化界面数据
     */
    private boolean isInitDataHDB = true;

    //============================================================
    //  开放配置
    //============================================================

    /**
     * 用于控制每个Fragment进入{@link BaseFragment#setUserVisibleHint(boolean)} 时，是否重新执行onRequest()方法
     */
    public boolean isLazyLoadHDB = true;
    /**
     * 是否Log打印Fragment的生命周期
     */
    public boolean isLogFragmentLife = false;

    //============================================================
    //  全局变量
    //============================================================

    /**
     * 手机屏幕宽度参数
     */
    public int screenWidth;
    /**
     * 手机屏幕高度参数
     */
    public int screenHeight;
    /**
     * 界面视图布局
     */
    public View rootLayout;

    public Context context;
    public Activity activity;
    public Application application;

    @Override
    public void onAttach(Context context) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onAttach(Context context)");
        }
        super.onAttach(context);

        this.context = context;
    }

    @Override
    public void onAttach(Activity activity) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onAttach(Activity activity)");
        }
        super.onAttach(activity);

        this.activity = activity;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onCreate(Bundle savedInstanceState)");
        }
        super.onCreate(savedInstanceState);

        this.application = activity.getApplication();
        this.screenWidth = ScreenUtils.getScreenWidth();
        this.screenHeight = ScreenUtils.getScreenHeight();

        EventBus.getDefault().register(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onViewCreated(View view, Bundle savedInstanceState)");
        }
        super.onViewCreated(view, savedInstanceState);

        this.isCreateed = true;
        if (rootLayout == null) {
            rootLayout = view;
        }
        /* 初始化界面数据 */
        if (isInitViewHDB) {
            initViewHDB(view, savedInstanceState);
            isInitViewHDB = false;
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onActivityCreated(Bundle savedInstanceState)");
        }
        super.onActivityCreated(savedInstanceState);

        if (isInitDataHDB) {
            initDataHDB(savedInstanceState);
            isInitDataHDB = false;
        }
    }

    @Override
    public void onStart() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onStart()");
        }
        super.onStart();
    }

    @Override
    public void onResume() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onResume()");
        }
        super.onResume();

        if (getUserVisibleHint() && isCreateed) {
            onRefreshHDB();
        }
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - setUserVisibleHint(" + isVisibleToUser + ")");
        }
        super.setUserVisibleHint(isVisibleToUser);

        if (isVisibleToUser && isCreateed) {
            onVisiableHDB();
            if (isLazyLoadHDB) {
                onLazyLoadHDB();
                isLazyLoadHDB = false;
            }
        }
    }

    @Override
    public void onPause() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onPause()");
        }
        super.onPause();
    }

    @Override
    public void onStop() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onStop()");
        }
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onDestroyView()");
        }
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onDestroy()");
        }
        onFinishing();

        EventBus.getDefault().unregister(this);

        this.isCreateed = false;
        super.onDestroy();
    }

    @Override
    public void onDetach() {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onDetach()");
        }
        super.onDetach();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (isLogFragmentLife) {
            LogUtils.d("Fragment - " + this.getClass().getSimpleName() + " - onActivityResult(" + requestCode + "," + resultCode + ") data.size = " + ((data == null || data.getExtras() == null) ? 0 : data.getExtras().size()));
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void initViewHDB(View view, @Nullable Bundle savedInstanceState) {
    }

    @Override
    public void initDataHDB(@Nullable Bundle savedInstanceState) {
    }

    @Override
    public void onRefreshHDB() {
    }

    @Override
    public void onLazyLoadHDB() {
    }

    @Override
    public void onVisiableHDB() {
    }

    @Override
    public void onFinishing() {
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void getMessageEventOnMain(BaseMessageEvent baseMessageEvent) {
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void getMessageEventOnBackground(BaseMessageEvent baseMessageEvent) {
    }
}

File: library/src/main/java/com/handy/basic/base/BaseFragment.kt
package com.handy.basic.base

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import com.blankj.utilcode.util.LogUtils
import com.blankj.utilcode.util.ScreenUtils
import com.handy.basic.eventbus.BaseEventMessage
import com.handy.basic.mvp.BasePresenter
import com.handy.basic.mvp.IView
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * @title: BaseFragment
 * @projectName HandyBasicKT
 * @description: Fragment通用超类
 * @author LiuJie https://github.com/Handy045
 * @date Created in 2019-10-11 16:29
 */
open class BaseFragment<A : BaseActivity<P>, P : BasePresenter> : Fragment(),
    BaseApplicationApi.FragmentApi, IView<P> {
    //============================================================
    //  私有配置
    //============================================================

    /**
     * 是否已创建
     */
    private var isCreated = false
    /**
     * 初始化界面视图
     */
    private var isInitializedView = true
    /**
     * 初始化界面数据
     */
    private var isInitializedData = true

    //============================================================
    //  开放配置
    //============================================================

    /**
     * 是否使用了新的生命周期启动模式
     */
    var isUseMaxLifecycle = false
    /**
     * 是否使用懒加载
     */
    var isUseLazyLoad = true
    /**
     * 是否在日志打印生命周期
     */
    var isPrintLifecycle = false
    /**
     * 屏幕宽度
     */
    var screenWidth = 0
    /**
     * 屏幕高度
     */
    var screenHeight = 0
    /**
     * 界面视图布局
     */
    lateinit var rootLayout: View
    /**
     * 所挂载的Activity
     */
    var parentActivity: A? = null
        get() {
            if (field == null) {
                field = this.mActivity as? A
            }
            return field
        }
    /**
     * MVP处理者
     */
    var presenter: P? = null
        get() {
            if (field == null) {
                field = initPresenter()
            }
            return field
        }
    /**
     * 保存的实例状态
     */
    var savedInstanceState: Bundle? = null
    /**
     * 上下文
     */
    lateinit var mContext: Context
    /**
     * 父类实例
     */
    lateinit var mActivity: Activity

    //============================================================
    //  方法区
    //============================================================

    override fun onAttach(context: Context) {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onAttach(context: Context)")
        }
        super.onAttach(context)

        this.mContext = context
    }

    override fun onAttach(activity: Activity) {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onAttach(activity: Activity)")
        }
        super.onAttach(activity)

        this.mActivity = activity
        this.parentActivity = activity as? A
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onCreate(savedInstanceState: Bundle?)")
        }
        super.onCreate(savedInstanceState)

        try {
            this.presenter = initPresenter()
            this.savedInstanceState = savedInstanceState
            this.screenWidth = ScreenUtils.getScreenWidth()
            this.screenHeight = ScreenUtils.getScreenHeight()

            EventBus.getDefault().register(this)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onViewCreated(view: View, savedInstanceState: Bundle?)")
        }
        super.onViewCreated(view, savedInstanceState)

        try {
            this.isCreated = true
            this.rootLayout = view

            if (this.isInitializedView) {
                initViewHDB(view, savedInstanceState)
                this.isInitializedView = false
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onActivityCreated(savedInstanceState: Bundle?)")
        }
        super.onActivityCreated(savedInstanceState)

        try {
            if (this.isInitializedData) {
                initDataHDB(savedInstanceState)
                this.isInitializedView = false
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onStart() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onStart()")
        }
        super.onStart()
    }

    override fun onResume() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onResume()")
        }
        super.onResume()

        try {
            if (this.isCreated) {
                onResumeHDB()
                if (this.isUseMaxLifecycle && this.isUseLazyLoad) {
                    onLazyLoadHDB()
                    this.isUseLazyLoad = false
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - setUserVisibleHint(isVisibleToUser: Boolean)")
        }
        super.setUserVisibleHint(isVisibleToUser)

        try {
            if (this.isCreated && isVisibleToUser) {
                onVisiableHDB()
                if (!this.isUseMaxLifecycle && this.isUseLazyLoad) {
                    onLazyLoadHDB()
                    this.isUseLazyLoad = false
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onPause() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onPause()")
        }
        super.onPause()
    }

    override fun onStop() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onStop()")
        }
        super.onStop()
    }

    override fun onDestroyView() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onDestroyView()")
        }
        super.onDestroyView()
    }

    override fun onDestroy() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onDestroy()")
        }
        super.onDestroy()

        try {
            this.isCreated = false
            onFinishing()
            EventBus.getDefault().unregister(this)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onDetach() {
        if (this.isPrintLifecycle) {
            LogUtils.d("lifecycle - ${javaClass.simpleName} - onDetach()")
        }
        super.onDetach()
    }

    override fun initViewHDB(view: View, savedInstanceState: Bundle?) {
    }

    override fun initDataHDB(savedInstanceState: Bundle?) {
    }

    override fun onResumeHDB() {
    }

    override fun onVisiableHDB() {
    }

    override fun onLazyLoadHDB() {
    }

    override fun onFinishing() {
    }

    override fun initPresenter(): P? {
        return parentActivity?.presenter
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun getMessageEventOnMain(event: BaseEventMessage) {
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    fun getMessageEventOnBackground(event: BaseEventMessage) {
    }
}